src/expr/scalar/mod.rs

```diff
@@ -325,7 +332,34 @@ impl ScalarExpr {
                 }
             }
             ScalarExpr::CallVariadic { func, exprs } => {
-                if exprs.iter().all(|e| e.is_literal()) {
+                if *func == VariadicFunc::Coalesce {
+                    // First throw away any literal nulls. These can never
+                    // affect the result.
+                    exprs.retain(|e| !e.is_literal_null());
+
+                    // Then find the first argument that is a literal or
+                    // non-nullable column. All arguments after this argument
+                    // will be ignored, so throw them away. This intentionally
+                    // throws away errors that can never happen.
+                    if let Some(i) = exprs
+                        .iter()
+                        .position(|e| e.is_literal() || !e.typ(&relation_type).nullable)
+                    {
+                        exprs.truncate(i + 1);
+                    }
```

Comment:
Won't this fire on literal Null?

Reply:
Bah, lines up above that didn't click for me. Never mind!

Reply:
Yeah it's subtle, right? I added a ton of test cases because I got the logic wrong on the first five tries. I wasn't able to come up with a better way of expressing this logic, but open to suggestions/refactors.